### PR TITLE
refactor(swingset): provideVatKeeper replaces get / allocate

### LINKED ISSUE #3281: add kernelKeeper.provideVatKeeper(), remove getVatKeeper/allocateVatKeeper
## What is the Problem Being Solved?

As [noted](https://github.com/Agoric/agoric-sdk/pull/2784#discussion_r647858037) during the #2784 "Vat Warehouse" PR review, the split between `kernelKeeper.allocateVatKeeper()` (which must be called exactly once per kernel process) and `kernelKeeper.getVatKeeper()` (which must be called *after* `allocateVatKeeper`) is awkward. The Vat Warehouse uses `kk.getVatKeeper() || kk.allocateVatKeeper()` as a pseudo "provide" function (where we use `provideFoo()` as our standard name for a function that either creates and returns a new Foo, or returns a previously-created Foo). However `vatTranslator.js` only calls `getVatKeeper()`, which makes it dependent upon someone else having called `allocate` first, which is fragile and likely to break under maintenance.

## Description of the Design

* add `kernelKeeper.provideVatKeeper()`, which can use the existing internal ephemeral `vatKeepers` table as a cache
* remove `kernelKeeper.getVatKeeper()` and `allocateVatKeeper()`
* replace all calls to either function with `provideVatKeeper()`

## Performance Considerations

This approach is fine for now, but as our system grows biggest (in particular as we get a larger number of short-lived vats), the internal ephemeral `vatKeepers` table will grow too much: it will contain a `vatKeeper` for every vat that was ever used during the lifetime of the kernel process, whereas we only need them for active vats (vats which are paged in by the Vat Warehouse).

One fix would be to add a `kernelKeeper.evictVatKeeper()` method, and have Vat Warehouse call it when evicting a vat from memory.

A different fix would be to have `kernelKeeper` apply an LRU policy to its internal cache, to drop some vatKeepers.

I'm currently in favor of the `evictVatKeeper` approach, as it would basically retain a 1:1 relationship between Vat Warehouse paged-in vats and `kernelKeeper` "paged-in" `vatKeeper` instances, without binding the two together very tightly (`evictVatKeeper` is more of a hint, since any call to `provideVatKeeper` will just make a new one).

It's worth remembering that `vatKeeper`s are just a convenience: they no longer contain any state in RAM, they just close over the `vatID` in a way that makes the API easier to use. If managing them becomes awkward, we could remove them in favor of a big batch of new `kernelKeeper` methods that all take a `vatID` as their first argument.

## Test Plan

Normal unit test in `test-state.js`.

### PR DESCRIPTION
**Pull request by dckc**
fixes #3281

### CODE REVIEW

**Comment by warner** [comment]
Looks pretty good, I think I'd change `hasRunningVat` to `isVatAlive` or `vatIsAlive` or something, and get rid of `evictVatKeeper` unless you think vatWarehouse is going to call it after paging out a vat or something.

**Comment by dckc** [comment]
> ... change `hasRunningVat` to ... `vatIsAlive`

done.

> .. and get rid of `evictVatKeeper` unless you think vatWarehouse is going to call it after paging out a vat or something.

vatWarehouse _does_ evict the VatKeeper ... er... tee hee... it does now.

_oops... I squashed those two fixups._

PTAL.

File: packages/SwingSet/src/kernel/state/kernelKeeper.js [hunk 1]
```diff
@@ -779,6 +777,28 @@ export default function makeKernelKeeper(kvStore, streamStore, kernelSlog) {
...
+   */
+  function closeVatTranscript(vatID) {
+    insistVatID(vatID);
+    const transcriptStream = `transcript-${vatID}`;
+    streamStore.closeStream(transcriptStream);
+  }
+
+  function evictVatKeeper(vatID) {
+    insistVatID(vatID);
+    closeVatTranscript(vatID);
```
Comment (warner) [inline packages/SwingSet/src/kernel/state/kernelKeeper.js]: `vatWarehouse.evict()` is calling both `closeVatTranscript()` and `evictVatKeeper()`, so I think either it should just call `evictVatKeeper`, or this call to `closeVatTranscript` could go away. Not a big deal to do both (`streamStore.closeStream` should ignore the second one), but slightly redundant.

**Review by warner — APPROVED** [review]
looks good, thanks for the updates!

**Pull request merged**